internal/plugin/sign.go

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright The Helm Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"sigs.k8s.io/yaml"
+
+	"helm.sh/helm/v4/pkg/provenance"
+)
+
+// SignPlugin signs a plugin using the SHA256 hash of the tarball.
+//
+// This is used when packaging and signing a plugin from a tarball file.
+// It creates a signature that includes the tarball hash and plugin metadata,
+// allowing verification of the original tarball later.
+func SignPlugin(tarballPath string, signer *provenance.Signatory) (string, error) {
+	// Extract plugin metadata from tarball
+	pluginMeta, err := ExtractPluginMetadata(tarballPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to extract plugin metadata: %w", err)
+	}
+
+	// Marshal plugin metadata to YAML bytes
+	metadataBytes, err := yaml.Marshal(pluginMeta)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal plugin metadata: %w", err)
+	}
+
+	// Use the generic provenance signing function
+	return signer.ClearSign(tarballPath, metadataBytes)
+}
+
+// ExtractPluginMetadata extracts plugin metadata from a tarball
+func ExtractPluginMetadata(tarballPath string) (*Metadata, error) {
+	f, err := os.Open(tarballPath)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	return ExtractPluginMetadataFromReader(f)
+}
+
+// ExtractPluginMetadataFromReader extracts plugin metadata from a tarball reader
+func ExtractPluginMetadataFromReader(r io.Reader) (*Metadata, error) {
```

Comment:
```suggestion
func ExtractTgzPluginMetadata(r io.Reader) (*Metadata, error) {
```

Reply:
cool to do this one in this PR?
- https://github.com/helm/helm/pull/31196

Reply:
added to #31196 :+1: i agree it's a more explicit name